### PR TITLE
Processing data in models 

### PR DESCRIPTION
**Pull request by diarisdiakite**
In this PR, I:
📃 Created models
✔️ Created model classes for all entities as shown in the [ERD diagram](https://github.com/diarisdiakite/Blog-app/images/blog_app_erd_v1_1.png).
✔️ Set up associations between models with the author_id column in posts table as the foreign_key for the users table.

📃 Used models to insert data
✔️ Created at least one user 
✔️ Created at least 4 posts written by one of the users I created
✔️ Created at least 6 posts comments for one of the posts I created
✔️ Use other [CRUD methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/active_record_crud.md) to find, update and delete entities.

📃 Created custom methods in the models
✔️ For Users : a method that returns the 3 most recent posts for a given user.
✔️ For Posts: a method that updates the posts counter for a user and a method which returns the 5 most recent comments for a given post.
✔️ For Comments: a method that updates the comments counter for a post.
✔️ For Likes: a method that updates the likes counter for a post.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,2 +1,26 @@
-﻿class User < ApplicationRecord
+class User < ApplicationRecord
+  has_many :posts, foreign_key: 'author_id'
```
Comment (topeogunleye) [inline app/models/user.rb:2]: Hello @diarisdiakite,

First off, I'd like to commend you for the effort you've put into setting up the relationships in your `user.rb` model. Your implementation for associating the `User` and `Post` models using the `author_id` foreign key is spot on and aligns perfectly with the ERD diagram. Well done!

However, I noticed a small oversight. While you have the `Post` relationship set up correctly, the associations for `comments` and `likes` seem to be missing. To align with the project requirements, you might want to add:

```ruby
has_many :comments
has_many :likes
```

By incorporating these changes, your `User` model will be comprehensive and in complete harmony with the project's expectations.

Keep up the fantastic work! Remember, coding is a journey, and every step, whether big or small, contributes to your growth.

Warm regards,
Temitope

File: app/models/post.rb [hunk 2]
```diff
@@ -1,3 +1,39 @@
...
+  # Create method that updates the comments counter for a post.
+  def update_comments_counter
+    update(comments_counter: comments.count)
+  end
+
+  # Create method that updates the likes counter for a post.
+  def update_likes_counter
+    update(likes_counter: likes.count)
+  end
+
```
Comment (topeogunleye) [inline app/models/post.rb:17]: Hello @diarisdiakite,

Great work on your `Post` model! The way you've managed comments and likes is efficient, and the "Read more" prompt for longer comments is a thoughtful touch.

However, the project requirements mention updating the `User` model's post counter. I suggest using Rails' `after_save` callback to automatically update this count whenever a post is added:

```ruby
# app/models/post.rb
after_save :update_post_counter

private

def update_post_counter
  author.increment!(:posts_counter)
end
```

This approach ensures the `User` model always has the accurate post count and encapsulates the logic within the `Post` model.

Your code showcases strong Rails knowledge. With this minor addition, you'll be perfectly aligned with the project guidelines.

Keep it up!
Best,
Temitope

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @diarisdiakite, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

Try-Again

> ## (Highlights) Good Points: 💪 

1. **Organized Directory Structure** 📁: You've kept the `User`, `Post`, `Comment`, and `Like` models neatly inside the `app/models/` directory, which follows Rails conventions and ensures maintainability.

2. **Correct Associations** 🔗: Your model associations are on point! The foreign key and class name attributes have been correctly used, making sure that the relationships between your models align with the given ERD diagram. This solid foundation will enable smooth interactions between models and help in building out further features seamlessly.

3. **Efficient Use of Active Record** 💾: The `has_many` and `belongs_to` methods have been employed optimally in your models. This will provide you with a plethora of methods to retrieve and manipulate data effortlessly.

4. **Effective Retrieval Methods** 🚀: Kudos for implementing methods in the `User` model to fetch the 3 most recent posts and in the `Post` model to retrieve the 5 most recent comments. Such methods enhance the user experience by providing timely and relevant information.

In summary, your attention to detail, adherence to Rails conventions, and effective use of ActiveRecord associations is commendable 🌟. It's evident that you have a good understanding of how models interact in a Rails application. Keep up the fantastic work!

## Required Changes ♻️
_Check the comments under the review._
### I couldn't use GitHub inline Comments because no changes were made to the comment and like model.

---

Hello,

I've reviewed your code, and you're making solid strides! I noticed a couple of areas where you can enhance the functionality further:

1. Within the `Comment` model, it appears that there's no method to update the comments counter for a post. This is crucial as it will keep the `Post` model up-to-date with the number of associated comments. 

2. Similarly, for the `Like` model, there isn't a method to update the likes counter for a post. Keeping an accurate count of likes is essential for user experience and for potential features you might want to implement in the future. 

Here's a brief example for the `Like` model:

```ruby
# app/models/like.rb
...

def update_likes_counter
  post.increment!(:likes_counter)
end
```

Implementing these methods will ensure both the `Post` model's comments and likes counters remain accurate, aligning your application more closely with the project requirements.

You're doing fantastic work, and with these additions, your project will be even more comprehensive!

Best wishes,
Temitope

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 2 above)
Comment (diarisdiakite) [reply app/models/post.rb:17]: Thank you very much @topeogunleye for these insightful comments, I've learnt a lot from them.

File: app/models/user.rb (hunk 1 above)
Comment (diarisdiakite) [reply app/models/user.rb:2]: done @topeogunleye. Thank you very much.

**Review by tugsem — APPROVED** [review]
## STATUS: APPROVED 👍🎯  
Hello @diarisdiakite 👋 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights ⭐ 
✔️ required changes are implemented
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 2 above)
Comment (topeogunleye) [reply app/models/post.rb:17]: you're welcome

**Pull request merged**